lib/internal/modules/cjs/loader.js

```diff
@@ -1004,18 +1004,26 @@ Module._resolveFilename = function(request, parent, isMain, options) {
     paths = Module._resolveLookupPaths(request, parent);
   }
 
-  if (parent && parent.filename) {
-    const filename = trySelf(parent.filename, isMain, request);
-    if (filename) {
-      const cacheKey = request + '\x00' +
-          (paths.length === 1 ? paths[0] : paths.join('\x00'));
-      Module._pathCache[cacheKey] = filename;
-      return filename;
-    }
+  let base, filename;
+  if (parent && (parent.filename || parent.id === '<repl>')) {
+    base = parent.filename || process.cwd() + path.sep;
+  } else if (parent && parent.id === 'internal/preload') {
+    try {
+      base = process.cwd() + path.sep;
+    } catch {}
+  }
+  if (base) {
+    filename = trySelf(base, isMain, request);
+  }
+  if (filename) {
+    const cacheKey = request + '\x00' +
+        (paths.length === 1 ? paths[0] : paths.join('\x00'));
+    Module._pathCache[cacheKey] = filename;
+    return filename;
   }
 
   // Look up the filename first, since that's the cache key.
-  const filename = Module._findPath(request, paths, isMain, false);
+  filename = Module._findPath(request, paths, isMain, false);
```

Comment:
This feels a bit confusing. Can we use a different name above? It took me a good second to figure out that the previous value of `filename` is completely ignored.